### PR TITLE
Use the user_login as user_id for OpenID

### PR DESCRIPTION
**Pull request by akirk**
We should not expose the internal WordPress user id but instead only the user login.

### CODE REVIEW

**Review by psrpinto — APPROVED** [review]
Looks good! Got some minor remarks.

**Comment by ashfame** [comment]
We shouldn't do this, as changing of usernames is a thing we entertain. If we are not comfortable with exposing internal WordPress user ID, we can just hash (sha256 or bcrypt but definitely not md5) it.

**Comment by psrpinto** [comment]
> We shouldn't do this, as changing of usernames is a thing we entertain.

Could you clarify this, I'm not sure I follow @ashfame. I was under the impression it was not possible to change user username, at least it doesn't seem to be possible through wp-admin.

>  we can just hash it

I don't think that would solve the problem as the hash would also change when the id changes.

----

I think relying on usernames instead of the WordPress integer id is a superior solution for multiple reasons, assuming it's not possible to change a user's username):

1. The fact that WordPress stores user ids as integers is an implementation detail that should not be exposed to external systems. If WordPress would start using random hashes as user ids (unlikely, I know) the link between the user and the OIDC server (Is that the correct term?) would be broken.
2. WordPress site administrators have no control over the auto-incrementing user ids, while they have some control over the username. The username can be manipulated in the database, the auto-incrementing id cannot.
3. Usernames are more explicit, it's easier to understand `psrpinto` than `7647844876`. Only useful for debugging.

**Comment by akirk** [comment]
> We shouldn't do this, as changing of usernames is a thing we entertain.

I would argue that then the Matrix username should change along with it, and since this is not possible it needs to be retired and the new one chosen.

File: class-oauth2-storage.php [hunk 1]
```diff
@@ -58,8 +58,8 @@ public static function validate_id_token( $id_token ) {
 		return self::validate_string_length( $id_token, 2000 );
 	}
 
-	public static function validate_user_id( $user_id ) {
-		return intval( $user_id );
+	public static function validate_user_login( $user_login ) {
+		return self::validate_string_length( $user_login, 60 );
```
[the remark this answers is not in the document]
Comment (akirk) [reply class-oauth2-storage.php:62]: Yes: https://github.com/WordPress/WordPress/blob/66d425d86aaabd3bb5def79d7eee13d53936122d/wp-admin/includes/schema.php#L193

This is actually for a `sanitize_callback`, I'll rename the functions. This is just to ensure that rules are enforced to the meta before storing it.

File: class-rest.php [hunk 2]
```diff
@@ -61,7 +61,13 @@ public function authorize() {
 			exit;
 		}
 
-		$this->server->handleAuthorizeRequest( $request, $response, true, get_current_user_id() );
+		$user = wp_get_current_user();
+		if ( ! $user ) {
+			$response->send();
```
[the remark this answers is not in the document]
Comment (akirk) [reply class-rest.php:66]: This is just defensive programming, this if should not be reachable since we check for `is_user_logged_in()` before.

**Review by psrpinto — APPROVED** [review]
🚀

**Comment by ashfame** [comment]
> Could you clarify this, I'm not sure I follow @ashfame. I was under the impression it was not possible to change user username, at least it doesn't seem to be possible through wp-admin.

@psrpinto Sure! It is indeed not possible to change username via UI as its best described as an external identifier for the user, but this is something we accept requests for, from at least Automatticians (not sure for general users on WPCOM).

But if we take a moment to think about why do we do this and simply not ask users to just create new accounts, it's probably for the data associated with their account. Changing the username which only affects how externally this account is lookup-able, sounds like a good option with no data loss if changing outside identifier is the exact need.

> I don't think that would solve the problem as the hash would also change when the id changes.

Perhaps I don't understand the problem here. Why do we need to support the change of user IDs? That's the internal identifier for a particular user and should never ever change. If this is just for supporting instances where we want to ensure things continue to work if somebody wipes their WP data and start over, then I don't feel this is a good measure. Is there a use case I am missing that requires supporting change of user IDs?

Looking at this PR alone, if we just want to avoid exposing the ID as a security caution, and all we care for is having an identifier that stays the same, hashing the ID does the job and we would never have to worry about updating it, even with username change requests.

> I would argue that then the Matrix username should change along with it, and since this is not possible it needs to be retired and the new one chosen.

@akirk With username change requests (which I would say are more likely to happen as opposed to change in user ID), and our Matrix account being tied up to username, that user's matrix access breaks. If SSO is tied up to ID or hash of the ID, then we have the flexibility of not breaking the link with matrix account or choose to retire the matrix account, whatever is desired/requested by the user. This is better than always having to break (need to retire and use a new account) the matrix account, even when not desired.

Hence, using `ID` or hash of it, something that will not change, seems like a better choice than username, which is [known to happen](https://mc.a8c.com/mgs/?q=username+change+request).

**Comment by psrpinto** [comment]
This came indeed from the use case of clearing the database: https://github.com/Automattic/Orbit/issues/312

I didn't consider that WP usernames could be changed, so my arguments from my previous comment don't really apply. Considering clearing the db is probably a weird use case (though we did encounter it immediately), I don't really have an opinion on this. Feels like it's a trade-off between two use cases:

1. Supporting changing usernames
2. Supporting clearing database

Another option could be to use email address, but that probably comes with its own set of challenges.

**Comment by ashfame** [comment]
While I appreciate the will to make the software more robust to handle clearing WP database as well, I think making a choice that leads to breaking matrix account access in future or being forced to retire existing account to use new account (when its not desirable) is a decision we should really think twice about.

The above need is probably handled with a prefix to the user ID and when a site administrator wants to avoid incorrect linking of accounts, they can change that prefix and no account will be mistakenly link. Every user would get a new account, which is probably justified if they are clearing their WP database or users table.

**Comment by akirk** [comment]
I'm going to merge this because I believe that it will lead to more confusion when you migrate between SSO systems that people might gain control over random coworkers/community members' Matrix accounts, than the issues caused by people changing their usernames. I'd argue that if they want to change their username on one system, they don't want to continue using that username on Matrix either.

**Comment by akirk** [comment]
Or explained otherwise: if I talk to a user `@xyz:community.wordpress.org` I need to be sure to be talking the same user as `https://profiles.wordpress.org/xyz`.

What about we add a filter so that it could be possible to override that to the old username so that in case someone do wants to keep that old user, the filter could still report the old username to Matrix.

**Pull request merged**